src/main/haskell/kore/src/Kore/Step/Representation/Semilattice.hs

```diff
@@ -0,0 +1,68 @@
+{-|
+Module      : Kore.Step.Representation.Semilattice
+Description : Data structures and functions for handling semilattice terms.
+Copyright   : (c) Runtime Verification, 2019
+License     : NCSA
+Maintainer  : virgil.serbanuta@runtimeverification.com
+Stability   : experimental
+Portability : portable
+
+This is intended to be imported qualified:
+
+@
+import           Kore.Step.Representation.Semilattice
+                 ( Semilattice )
+import qualified Kore.Step.Representation.Semilattice as Semilattice
+@
+-}
+module Kore.Step.Representation.Semilattice
+    ( ElementType (..)
+    , Semilattice (..)
+    ) where
+
+import           Data.Proxy
+                 ( Proxy (..) )
+import qualified Data.Set as Set
+
+{-| 'ElementType' identifies element types in a semilattice.
+-}
+data ElementType
+  = Absorbing
+  -- ^ An element a such that ∀x . x * a = a
+  | Neutral
+  -- ^ An element e such that ∀x . x * e = x
+  | Other
+  -- ^ Any element which is not identified by another ElementType case.
+
+{-| A semilattice with an absorbing and neutral elements.
+
+The absorbing and neutral elements are assumed to be different (i.e. the
+semilattice should have more than one element).
+-}
+class Ord element => Semilattice term element where
```

Comment:
I feel a little weird about tying `term` and `element` together through the `Semilattice` class. Usually we would call `element` by itself a (join or meet) semilattice. The `term` is not really a part of the semilattice structure of `element`, it's just a container that respects the semilattice structure of `element`. It's also a little weird to me that the `Semilattice` class doesn't know if it's a join or meet semilattice. How would you feel about the following instead?

~~~ haskell
-- | Semigroup operation respects the meet (@\\and@) semilattice of @child@.
instance TopBottom child => Semigroup (MultiAnd child) where
    ...

-- | Identity is @\\top@.
instance TopBottom child => Monoid (MultiAnd child) where
    ...
~~~

...and likewise for `MultiOr`. The `Semigroup` definition of `(<>)` would both merge and simplify. We would need to add functions to `TopBottom` for constructing the `top` and `bottom` of each type. (Then we would be able to write laws for that class!) The implementation of `simplify` could still be a single function, just specialized in each module for `MultiAnd` and `MultiOr`:

~~~ haskell
import GHC.Exts (IsList, fromList, toList)

simplifySemilattice :: IsList term => (element -> ElementType) -> term element -> term element
simplifySemilattice elementType =
    fromList . simplifyIdempotency . simplifyAbsorbingNeutral . toList
  where
    simplifyIdempotency :: [element] -> [element]
    simplifyIdempotency = Set.toList . Set.fromList
    simplifyAbsorbingNeutral :: [element] -> [element]
    simplifyAbsorbingNeutral elements = go [] elements
      where
        go filtered [] = reverse filtered
        go filtered (element:unfiltered) =
            case elementType element of
                Absorbing -> [element]
                Neutral -> go filtered unfiltered
                Other -> go (element:filtered) unfiltered

joinSemilattice :: TopBottom element => element -> ElementType
joinSemilattice e
  | isTop e    = Absorbing
  | isBottom e = Neutral
  | otherwise  = Other

meetSemilattice :: TopBottom element => element -> ElementType
meetSemilattice e
  | isTop e    = Neutral
  | isBottom e = Absorbing
  | otherwise  = Other

simplifyMultiAnd = simplifySemilattice meetSemilattice

simplifyMultiOr = simplifySemilattice joinSemilattice
~~~